thrift/thrift-gen/tchannel-template.go

```diff
@@ -142,46 +149,57 @@ func (s *{{ .ServerStruct }}) Handle(ctx {{ contextType }}, methodName string, p
 }
 
 {{ range .Methods }}
-	func (s *{{ $svc.ServerStruct }}) {{ .HandleFunc }}(ctx {{ contextType }}, protocol athrift.TProtocol) (bool, athrift.TStruct, error) {
+	func (s *{{ $svc.ServerStruct }}) {{ .HandleFunc }}(ctx {{ contextType }}, protocol athrift.TProtocol) (handled bool, resp athrift.TStruct, err error) {
 		var req {{ .ArgsType }}
 		var res {{ .ResultType }}
+		const serviceMethod = "{{ $svc.ThriftName }}::{{ .ThriftName }}"
 
-		if err := req.Read(protocol); err != nil {
-			return false, nil, err
+		if readErr := req.Read(protocol); readErr != nil {
+			return false, nil, readErr
 		}
 
-		{{ if .HasReturn }}
-			r, err :=
-		{{ else }}
-			err :=
-		{{ end }}
-				s.handler.{{ .Name }}({{ .CallList "req" }})
+		postRun, err := s.interceptorRunner.RunPre(ctx, serviceMethod, &req)
 
-		if err != nil {
-			{{ if .HasExceptions }}
-			switch v := err.(type) {
-				{{ range .Exceptions }}
+		defer func () {
+			err = postRun(resp, err)
+			if err != nil {
+				resp = nil
+				{{ if .HasExceptions }}
+				switch v := err.(type) {
+					{{ range .Exceptions }}
 					case {{ .ArgType }}:
 						if v == nil {
-							return false, nil, fmt.Errorf("Handler for {{ .Name }} returned non-nil error type {{ .ArgType }} but nil value")
+							err = fmt.Errorf("Handler for {{ .Name }} returned non-nil error type {{ .ArgType }} but nil value")
+						} else {
+							res.{{ .ArgStructName }} = v
+							err = nil
+							resp = &res
 						}
-						res.{{ .ArgStructName }} = v
+					{{ end }}
+				}
 				{{ end }}
-					default:
-						return false, nil, err
 			}
-			{{ else }}
-				return false, nil, err
-			{{ end }}
-		} else {
-    {{ if .HasReturn }}
-		  res.Success = {{ .WrapResult "r" }}
+		}()
```

Comment:
i don't think `handled` is being set correctly in the case where a post interceptor returns a non-thrift error?